engine/src/main/java/net/jqwik/engine/support/types/TypeUsageImpl.java

```diff
@@ -681,18 +641,83 @@ public Optional<TypeUsage> getSuperclass() {
 		if (rawType.getSuperclass() == null) {
 			return Optional.empty();
 		}
-		return Optional.of(TypeUsage.forType(rawType.getGenericSuperclass()));
+		// Double checking for thread safety
+		if (superclass == null) {
+			synchronized (this) {
+				if (superclass == null) {
+					superclass = createSuperclass();
+				}
+			}
+			superclass = createSuperclass();
+		}
+		return Optional.of(superclass);
+	}
+
+	private TypeUsageImpl createSuperclass() {
+		TypeUsageImpl superclass1 = TypeUsageImpl.forType(rawType.getGenericSuperclass());
+		superclass1.replaceTypeVariablesFromSubtype(this);
+		return superclass1;
+	}
+
+	private TypeUsageImpl replaceTypeVariablesFromSubtype(TypeUsageImpl subtype) {
+		Map<String, TypeUsage> typeVariables = extractTypeVariables(subtype);
+		replaceTypeVariablesIn(typeVariables, this);
+		return this;
+	}
+
+	private static void replaceTypeVariablesIn(Map<String, TypeUsage> typeVariables, TypeUsageImpl replacementTarget) {
+		for (int i = 0; i < replacementTarget.typeArguments.size(); i++) {
+			TypeUsage typeArgument = replacementTarget.typeArguments.get(i);
+			if (typeArgument.isTypeVariable() && typeVariables.containsKey(typeArgument.getTypeVariable())) {
+				replacementTarget.typeArguments.set(i, typeVariables.get(typeArgument.getTypeVariable()));
+			} else {
+				replaceTypeVariablesIn(typeVariables, (TypeUsageImpl) typeArgument);
+			}
+		}
+	}
+
+	@SuppressWarnings("rawtypes")
+	private static Map<String, TypeUsage> extractTypeVariables(TypeUsage subtype) {
+		List<TypeUsage> typeArgumentsFromSubtype = subtype.getTypeArguments();
+		Map<String, TypeUsage> namedTypeArguments = new HashMap<>();
+		for (int pos = 0; pos < typeArgumentsFromSubtype.size(); pos++) {
+			TypeUsage typeArgumentFromSubtype = typeArgumentsFromSubtype.get(pos);
+			if (!typeArgumentFromSubtype.isTypeVariable()) {
+				// The parameter name is taken from the raw type, e.g. List<String> -> E = String because List<E>
+				TypeVariable[] rawTypeArguments = subtype.getRawType().getTypeParameters();
+				if (rawTypeArguments.length >= pos) {
+					namedTypeArguments.put(rawTypeArguments[pos].getName(), typeArgumentFromSubtype);
+				}
+			}
+		}
+		return namedTypeArguments;
 	}
 
 	@Override
 	public List<TypeUsage> getInterfaces() {
-		return toTypeUsages(getRawType().getGenericInterfaces());
+		// Double checking for thread safety
+		if (interfaces == null) {
+			synchronized(this) {
```

Comment:
This is not safe, see https://shipilev.net/blog/2014/safe-public-construction/#_safe_initialization